subsys/emds/emds.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <emds/emds.h>
+
+#include <zephyr.h>
+#include <storage/flash_map.h>
+#include <drivers/flash.h>
+#include "emds_flash.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(emds, CONFIG_EMDS_LOG_LEVEL);
+
+K_SEM_DEFINE(emds_sem, 0, 1);
+static bool emds_ready;
+static bool emds_initialized;
+static uint32_t store_key;
+
+static sys_slist_t emds_dynamic_entries;
+static struct emds_fs emds_flash;
+static emds_store_cb_t app_store_cb;
+
+static void emds_handler(void)
+{
+	while (true) {
+		k_sem_reset(&emds_sem);
+		k_sem_take(&emds_sem, K_FOREVER);
+
+		k_sched_lock();
+
+		LOG_DBG("Emergency Data Storeage released");
+
+		STRUCT_SECTION_FOREACH(emds_entry, ch) {
+			ssize_t len = emds_flash_write(&emds_flash,
+						       ch->id, ch->data, ch->len);
+			if (len < 0) {
+				LOG_ERR("Write static entry: (%d) error (%d)",
+					ch->id, len);
+			} else if (len != ch->len) {
+				LOG_ERR("Write static entry: (%d) failed (%d:%d)",
+					ch->id, ch->len, len);
+			}
+		}
+
+		struct emds_dynamic_entry *ch;
+
+		SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
+			ssize_t len = emds_flash_write(&emds_flash,
+						       ch->entry.id, ch->entry.data, ch->entry.len);
+			if (len < 0) {
+				LOG_ERR("Write dynamic entry: (%d) error (%d).",
+					ch->entry.id, len);
+			}
+			if (len != ch->entry.len) {
+				LOG_ERR("Write dynamic entry: (%d) failed (%d:%d).",
+					ch->entry.id, ch->entry.len, len);
+			}
+		}
+
+		emds_ready = false;
+
+		k_sched_unlock();
+
+		/* Unlock all interrupts */
+		irq_unlock(store_key);
+
+		if (app_store_cb) {
```

Comment:
I would rather keep where it is. As then the k_sched_unlock and the irq_unlock is kept just as long as the storing is being done. That is critical. What the user does in the callback is then up to the user, but in that time the scheduler or interrupts does not need to be locked anymore.

Reply:
Ok. The problem with this, the user callback may get skipped from execution if the power runs out before that due to something else more time-consuming operation getting executed from the scheduler. 

If you want to keep this as is, at least this can be explained to users of callback. Can you update `app_store_cb` description stating something like this?
'Scheduler and interrupts are unlocked before triggering this callback. Therefore, this callback may not be reached before available backup power runs out.'

Reply:
Added your suggestion to the description.